### ISSUE TITLE
ribotin aborted with error 'std::bad_alloc'

### ISSUE THREAD

**Issue by zgliwen** [maickrau/ribotin#8]
Hello Mikko, 

Thanks for sharing ribotin, it is useful to extract the rDNA consensus sequence.

For most of my samples, ribotin runs well and quickly except for one sample. It aborts and gives the information "terminate called after throwing an instance of 'std::bad_alloc'" in the log file. 

I found that ribotin runs OK to extract rDNA-specific reads, but when getting consensus, the memory would be up to 50.9% and then stopped. Could you please give me some suggestions about how to resolve this problem?

 
The command I used is: 
ribotin/bin/ribotin-ref -r ref_seq.fa --orient-by-reference orient_seq.fa --mbg /data/miniconda3/envs/py3/bin/MBG -i input.fa -o outdir

The whole log file is:
ribotin-ref version Branch master commit d24db06ac2d219de846c5afea725d1c0a6713247 2023-03-06 16:21:10 +0200
output folder: /data/rDNA_unit/results/hifi_ribotin
extracting reads
running
running MBG
MBG command:
/data/miniconda3/envs/py3/bin/MBG -o /data/rDNA_unit/results/hifi_ribotin/graph.gfa -i /data/rDNA_unit/results/hifi_ribotin/reads.fa -k 101 -w 71 -a 2 -u 3 -r 15000 -R 4000 --error-masking=msat --output-sequence-paths /data/rDNA_unit/results/hifi_ribotin/paths.gaf --only-local-resolve 1> /data/rDNA_unit/results/hifi_ribotin/mbg_stdout.txt 2> /data/rDNA_unit/results/hifi_ribotin/mbg_stderr.txt
reading graph
getting consensus
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc

The file of rDNA-specific reads is about 68 MB.

 
Best regards,
Wen

**Comment by maickrau** [comment]
Sounds like it might be a bug. Could you share `outdir`?

**Comment by zgliwen** [comment]
Dear Mikko,

Thanks for your quick reply. Please download the output folder from https://www.dropbox.com/s/njxg9nluuf4dxg4/hifi_ribotin.rar?dl=0

Best,
Wen

**Comment by maickrau** [comment]
Could you also share ref_seq.fa and orient_seq.fa?

**Comment by zgliwen** [comment]
> Could you also share ref_seq.fa and orient_seq.fa?

Sure. Please download these two files from 
https://www.dropbox.com/scl/fo/1d8woahfctqe4i3tfq80c/h?dl=0&rlkey=8msq59eaia7pvvehz1ctrdto3

**Comment by maickrau** [comment]
This was caused by MBG trying to resolve the graph too much. You can rerun with the latest commit and with the parameter `--approx-morphsize 10000`

**Comment by zgliwen** [comment]
Thanks, Mikko. I am sure it will work.

But I am not sure what you mean by "rerun with the latest commit". Should I replace the old src folder with the latest one? Or should I clone the git and compile it again?

**Comment by maickrau** [comment]
I mean `git pull`, `make clean`, `make all` and then rerun with the newly compiled ribotin.

**Comment by zgliwen** [comment]
Hi, Mikko.

It said that there is no parameter "approx-morphsize". Is something wrong? 

The command is:
ribotin/bin/ribotin-ref -r ref_seq.fa --orient-by-reference orient_seq.fa --mbg /data/miniconda3/envs/py3/bin/MBG -i input.fa **--approx-morphsize 10000** -o outdir

The log is:
ribotin-ref version Branch master commit d24db06ac2d219de846c5afea725d1c0a6713247 2023-03-06 16:21:10 +0200
terminate called after throwing an instance of 'cxxopts::exceptions::no_such_option'
  what():  Option ‘approx-morphsize’ does not exist

**Comment by zgliwen** [comment]
Hi, Mikko.

I realized that yesterday I did not update it successfully.  I try again and it runs perfectly, thank you so much!

But I am curious about the parameter "--approx-morphsize". It sets the approximate length of one morph, so what is a morph? I wander to know in which situation I should consider this parameter. Thank you.

**Comment by maickrau** [comment]
It's the size of one copy of the rDNA repeat. So for example human about 45000, arabidopsis about 10000. The default is 45000 so if you're running it on something with smaller size like plants you should use the parameter. It's not important to get the exact size, probably any difference less than 5000 doesn't matter.

**Comment by zgliwen** [comment]
Hi, Mikko.

About two weeks ago, I ran ribotin but failed. So, I subsampled my data and ran it again with the default parameters. It is interesting that when I set --approx-morphsize to 10000 or 20000, the consensus sequences are different but nearly identical to those results from two different levels of subsampling separately. So I thought this parameter is something about subsampling at first. If it is just related to the size of an rDNA repeat, I should consider where the differences come from.

Thank you again for your help!
